From 756e579b8213d28edc8d0840295ea24a0c6880ac Mon Sep 17 00:00:00 2001
From: Erik Schumacher <erik.schumacher@iris-sensing.com>
Date: Tue, 19 Mar 2024 16:48:51 +0000
Subject: [PATCH] i2c-imx: Introduce optional busy polling for xfer

Mode can be selected in the device tree. Due to an unknown source
somewhere in the interrupt and wait queue handling, a delay of 300 us (or
multiples thereof) is introduced between the ACK bit from the peripheral
device and the transmission of the next bits. This is not a problem for
sporadic transmissions, but can accumulate in cases with heavy i2c
traffic. Reduce the udelay in i2c_imx_bus_busy() from 100 to 5 us.

imx8mp-irma6r2.dts: Use busy polling on i2c-1 ("epc bus")

Signed-off-by: Erik Schumacher <erik.schumacher@iris-sensing.com>
Signed-off-by: Jasper Orschulko <jasper@fancydomain.eu>
---
 arch/arm64/boot/dts/freescale/imx8mp-irma6r2.dts |  1 +
 drivers/i2c/busses/i2c-imx.c                     | 14 +++++++++++++-
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/freescale/imx8mp-irma6r2.dts b/arch/arm64/boot/dts/freescale/imx8mp-irma6r2.dts
index fd4577587358..58584030b0ee 100644
--- a/arch/arm64/boot/dts/freescale/imx8mp-irma6r2.dts
+++ b/arch/arm64/boot/dts/freescale/imx8mp-irma6r2.dts
@@ -179,6 +179,7 @@ &i2c2 {
 	clock-frequency = <384000>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_i2c2>;
+	force-atomic-rw;
 	status = "okay";
 
 	epc660_mipi: epc660_mipi@22 {
diff --git a/drivers/i2c/busses/i2c-imx.c b/drivers/i2c/busses/i2c-imx.c
index ed0ace73de51..162639ded5f5 100644
--- a/drivers/i2c/busses/i2c-imx.c
+++ b/drivers/i2c/busses/i2c-imx.c
@@ -287,6 +287,8 @@ struct imx_i2c_struct {
 	/* For checking slave events. */
 	spinlock_t     slave_lock;
 	struct hrtimer slave_timer;
+
+	bool force_atomic_rw;
 };
 
 static const struct imx_i2c_hwdata imx1_i2c_hwdata = {
@@ -591,7 +593,7 @@ static int i2c_imx_bus_busy(struct imx_i2c_struct *i2c_imx, int for_busy, bool a
 			return -ETIMEDOUT;
 		}
 		if (atomic)
-			udelay(100);
+			udelay(5);
 		else
 			schedule();
 	}
@@ -1367,6 +1369,13 @@ static int i2c_imx_xfer_common(struct i2c_adapter *adapter,
 	struct imx_i2c_struct *i2c_imx = i2c_get_adapdata(adapter);
 	int use_dma = 0;
 
+	/*
+	 * Some devices require polling transfers. Atomic transfers already
+	 * disable controller interrupts and bypass the DMA transfer paths.
+	 */
+	if (!atomic && i2c_imx->force_atomic_rw)
+		atomic = true;
+
 	/* Start I2C transfer */
 	result = i2c_imx_start(i2c_imx, atomic);
 	if (result) {
@@ -1809,6 +1818,9 @@ static int i2c_imx_probe(struct platform_device *pdev)
 		goto clk_notifier_unregister;
 	}
 
+	i2c_imx->force_atomic_rw = of_property_read_bool(pdev->dev.of_node,
+							 "force-atomic-rw");
+
 	/*
 	 * This limit caused by an i.MX7D hardware issue(e7805 in Errata).
 	 * If there is no limit, when the bitrate set up to 400KHz, it will
-- 
2.55.0

